eng/devices/android.cake

```diff
@@ -17,183 +18,190 @@ var TEST_RESULTS = Argument("results", EnvironmentVariable("ANDROID_TEST_RESULTS
-                arch = "arm64-v8a";
-        }
-        DEVICE_ID = $"system-images;android-{api};google_apis_playstore;{arch}";
-
-        // we are not using a virtual device, so quit
-        if (!emulator)
-            return;
-    }
-
-    Information("Test Device: {0}", TEST_DEVICE);
-    Information("Test Device ID: {0}", DEVICE_ID);
-
-    // delete the AVD first, if it exists
-    Information("Deleting AVD if exists: {0}...", ANDROID_AVD);
-    try { AndroidAvdDelete(ANDROID_AVD, avdSettings); }
-    catch { }
-
-    // create the new AVD
-    Information("Creating AVD: {0}...", ANDROID_AVD);
-    AndroidAvdCreate(ANDROID_AVD, DEVICE_ID, DEVICE_NAME, force: true, settings: avdSettings);
-
-    // start the emulator
-    Information("Starting Emulator: {0}...", ANDROID_AVD);
-    emulatorProcess = AndroidEmulatorStart(ANDROID_AVD, emuSettings);
-
-    // var waited = 0;
-    // while (AdbShell("getprop sys.boot_completed", adbSettings).FirstOrDefault() != "1") {
-    //     System.Threading.Thread.Sleep(1000);
-    //     if (waited++ > 60 * 10)
-    //         break;
-    // }
-    // Information("Waited {0} seconds for the emulator to boot up.", waited);
+	Information("Test Device: {0}", TEST_DEVICE);
+
+	// determine the device characteristics
+	{
+		var working = TEST_DEVICE.Trim().ToLower();
+		var emulator = true;
+		var api = 30;
+		// version
+		if (working.IndexOf("_") is int idx && idx > 0) {
+			api = int.Parse(working.Substring(idx + 1));
+			working = working.Substring(0, idx);
+		}
+		var parts = working.Split('-');
+		// os
+		if (parts[0] != "android")
+			throw new Exception("Unexpected platform (expected: android) in device: " + TEST_DEVICE);
+		// device/emulator
+		if (parts[1] == "device")
+			emulator = false;
+		else if (parts[1] != "emulator" && parts[1] != "simulator")
+			throw new Exception("Unexpected device type (expected: device|emulator) in device: " + TEST_DEVICE);
+		// arch/bits
+		if (parts[2] == "32") {
+			if (emulator)
+				DEVICE_ARCH = "x86";
+			else
+				DEVICE_ARCH = "armeabi-v7a";
+		} else if (parts[2] == "64") {
+			if (emulator)
+				DEVICE_ARCH = "x86_64";
+			else
+				DEVICE_ARCH = "arm64-v8a";
+		}
+		DEVICE_ID = $"system-images;android-{api};google_apis_playstore;{DEVICE_ARCH}";
+
+		// we are not using a virtual device, so quit
+		if (!emulator)
+			return;
+	}
+
+	Information("Test Device ID: {0}", DEVICE_ID);
+
+	// delete the AVD first, if it exists
+	Information("Deleting AVD if exists: {0}...", ANDROID_AVD);
+	try { AndroidAvdDelete(ANDROID_AVD, avdSettings); }
+	catch { }
+
+	// create the new AVD
+	Information("Creating AVD: {0}...", ANDROID_AVD);
+	AndroidAvdCreate(ANDROID_AVD, DEVICE_ID, DEVICE_NAME, force: true, settings: avdSettings);
+
+	// start the emulator
+	Information("Starting Emulator: {0}...", ANDROID_AVD);
+	emulatorProcess = AndroidEmulatorStart(ANDROID_AVD, emuSettings);
+
+	// wait for it to finish booting (10 mins)
+	var waited = 0;
+	var total = 60 * 10;
```

Comment:
When running on the agents, I have typically seen 5 minutes because the emulators are not accelerated. But, the tests are pretty fast. Just startup time is loooooooong.